# Project 2 - Market Intelligence Project: Data Integration & Visualization for Real Estate 🏠🏢

### **Objective:** Perform data cleaning, transformation, and integration from raw CSV files into a MySQL database for further business analysis. 

### **Stack:** Python (Pandas), SQL (MySQL), Power BI.

### 🏢 Business Context: 
Insight Places is a fictional property management company focused on short-term stays. As part of their market expansion strategy, the company needs to consolidate data from various sources to gain insights into pricing, customer satisfaction, and operational growth.

### 📂 Data Sources
This project utilizes six raw datasets provided in CSV format. Each file was processed and mapped from its original Portuguese structure to a standardized English relational database.

* 1-enderecos.csv (Addresses): Geographical information for all properties.

* 2-clientes.csv (Customers): Profile and contact information for guests.

* 3-proprietarios.csv (Owners): Details about property owners.

* 4-hospedagens.csv (Listings): Internal IDs, property types, and operational status.

* 5-alugueis.csv (Bookings): Internal IDs, booking dates, and total prices for stays.

* 6-avaliacoes.csv (Reviews): Customer feedback and property ratings.

### 1) Installing and Importing the libraries 

In this stage, I will load the libraries that I will use all over the project.

* **pandas:**  Manipulations and data analysis using dataframes.<br>
* **sqlalchemy:** A SQL toolkit that acts as a translator between Python objects and SQL databases, managing the connection bridge.<br> 
* **pymysql:** It provides the necessary interface for Python to send and receive data from MySQL.

In [2]:
#!pip install sqlalchemy
#!pip install pymysql

import pandas as pd
from sqlalchemy import create_engine
import pymysql 


#### 1.1) Installing data protection library

I will install python-dotenv to ensure password security.

In [5]:
#!pip install python-dotenv

#### 1.2) Environment Configuration and Security Validation

This step ensures that database credentials are securely loaded from environment variables, preventing sensitive information from being exposed in the source code.

In [3]:
import os
from dotenv import load_dotenv

# Load variables from the .env file
load_dotenv()

# Assigning credentials to variables using os-getenv
user = os.getenv('DB_USER')
password = os.getenv('DB_PASSWORD')
host = os.getenv('DB_HOST')
database = os.getenv('DB_NAME')

# Security Check: Validating if the environment variables are present
# I only check if the variables are NOT empty (None)
if all([user, password, host, database]):
    print("✅ Environment variables loaded successfully.")
    print("✅ System ready for database connection.")
else:
    print("❌ Error: One or more credentials are missing in the .env file.")

✅ Environment variables loaded successfully.
✅ System ready for database connection.


#### 1.3) Database Engine Initialization
With the credentials validated, I use SQLAlchemy to create the engine. This object will be the communication bridge between Python and MySQL, allowing for efficient data transfer.

In [4]:
# Using the variables that have already been loaded and validated in cell 1.2.
engine = create_engine("mysql+pymysql://{}:{}@{}/{}".format(user, password, host, database))
print("🚀 Connection engine is ready!")

🚀 Connection engine is ready!


### 2) Database Connection and Data Loading


⚙️ Technical Note: Handling Different CSV Separators
In Python's Pandas library, the read_csv() function defaults to a comma (,) as a separator. However, depending on the region (like in Brazil) or how the file was exported (e.g., from Excel), the separator might be a semicolon (;).

* Standard: pd.read_csv(path) (assumes comma)

* Custom: pd.read_csv(path, sep=';') (for semicolon-separated files)

**Why it matters:** If the wrong separator is used, Pandas will load the entire row as a single column, causing errors during data cleaning or database migration.

#### 2.1) Testing the CSV file reading with UTF-8 encoding

In [21]:
#Testing the CSV file reading using UTF-8 encoding
path = r"C:\Users\vanes\OneDrive\Área de Trabalho\PROJETOS EM ANDAMENTO\PORTFÓLIO\PROJETO 2\Parte_2\csv\1-enderecos.csv"  
df_addresses = pd.read_csv(path, encoding='utf-8')

# Displaying the first few rows to verify the encoding and data integrity
df_addresses.head()


,endereco_id,rua,numero,bairro,cidade,estado,cep
0,1,Lagoa de Teixeira,72,Tirol,Moraes,SP,87362-365
1,10,Colônia Davi Rodrigues,46,Saudade,Cunha,MS,45541-255
2,100,Núcleo Vitor Duarte,569,Lindéia,da Luz,TO,77228-924
3,1000,Loteamento Benício Gomes,65,Ernesto Nascimento,Nogueira do Oeste,AM,77002-693
4,10000,Vale Ana Julia Alves,30,Jardim Do Vale,Silveira,SP,95824-266


#### 2.2) Processing Table: Addresses 



In this section, I will establish a connection with the MySQL server to perform the ETL process, migrating the raw data from the 1-enderecos.csv file into the MySQL database for the `Addresses` table. I will rename the columns to match the English schema defined in the project.

In [ ]:
# --- TABLE 1: ADDRESSES  ---

# 1. Path
path = r"C:\Users\vanes\OneDrive\Área de Trabalho\PROJETOS EM ANDAMENTO\PORTFÓLIO\PROJETO 2\Parte_2\csv\1-enderecos.csv"

# 2. Load the CSV
df_addresses = pd.read_csv(path, encoding='utf-8')


# 3. Renaming columns to match our English Database Schema
df_addresses.columns = ['address_id', 'street', 'house_number', 'neighborhood', 'city', 'state', 'zip_code']

# 4. Migrating to MySQL
#try:
   # df_addresses.to_sql('addresses', con=engine, if_exists='append', index=False)
    #print("✅ Data successfully migrated to MySQL!")
#except Exception as e:
    #print(f"❌ Error: {e}")


✅ Data successfully migrated to MySQL!


#### 2.3) Processing Table: Customers 
In this section, I will perform the ETL process for the `Customers` table, migrating the raw data from the 2-clientes.csv file into the MySQL database. I will rename the columns to match the English schema established in the project. 

In [ ]:
# --- TABLE 2: CUSTOMERS  ---

# 1. Path 
path_customers = r"C:\Users\vanes\OneDrive\Área de Trabalho\PROJETOS EM ANDAMENTO\PORTFÓLIO\PROJETO 2\Parte_2\csv\2-clientes.csv" 

# 2. Load the CSV
df_customers = pd.read_csv(path_customers, encoding='utf-8')

# 3. Renaming columns to match our English Database Schema

df_customers.columns = ['customer_id', 'customer_name', 'customer_tax_id', 'contact']

# 4. Removing duplicates in the dataset
df_customers = df_customers.drop_duplicates(subset=['customer_tax_id'])

# 5. Migrating to MySQL

#try:
    #df_customers.to_sql('customers', con=engine, if_exists='append', index=False)
    #print("✅ Table 'customers' migrated successfully!")
#except Exception as e:
    #print(f"❌ Error migrating 'clients': {e}")


✅ Table 'customers' migrated successfully!


#### 2.4) Processing Table: Owners
In this section, I perform the ETL process for the `Owners` table. This step involves loading the raw data from the 3-proprietarios.csv file, renaming the columns to align with the database schema, and applying a data cleaning step to remove any duplicate records based on the tax_id. Although the database table includes a total_listings column, it is configured with a default value of 0, allowing for a seamless migration of the core owner information.

In [ ]:
# --- TABLE 3: OWNERS  ---

# 1. Path
path_owners = r"C:\Users\vanes\OneDrive\Área de Trabalho\PROJETOS EM ANDAMENTO\PORTFÓLIO\PROJETO 2\Parte_2\csv\3-proprietarios.csv"

# 2. Load the CSV 
df_owners = pd.read_csv(path_owners, encoding='utf-8')

# 3. Renaming to English Schema
# Verifique se o seu CSV tem: id, nome, cpf, contato
df_owners.columns = ['owner_id', 'owner_name', 'tax_id', 'contact']

# 4. Data Cleaning (Just in case there are duplicates here too!)
df_owners = df_owners.drop_duplicates(subset=['tax_id'])

# 5. Migration to MySQL
# try:
#     # The 'total_listings' column will be automatically filled with 0 by MySQL
#     df_owners.to_sql('owners', con=engine, if_exists='append', index=False)
#     print("✅ Table Owners migrated successfully!")
# except Exception as e:
#     print(f"❌ Error migrating 'owners': {e}")


✅ Table Owners migrated successfully!


#### 2.5) Processing Table: Listings

In this section, I process the `Listings` table, which is the core of the database. Since this table relies on Foreign Keys from both `Addresses` and `Owners`, the process includes rigorous data integrity checks and ID normalization to ensure a perfect migration.

#### 2.5.1)🔍 Data Integrity Check: Orphaned Records Discovery

In [18]:
# Checking if all address IDs in the Listings CSV exist in the 'addresses' table
db_addresses = pd.read_sql("SELECT address_id FROM addresses", con=engine)
missing_addresses = df_listings[~df_listings['address_id'].isin(db_addresses['address_id'])]

# Checking if all owner IDs in the Listings CSV exist in the 'owners' table
db_owners = pd.read_sql("SELECT owner_id FROM owners", con=engine)
missing_owners = df_listings[~df_listings['owner_id'].isin(db_owners['owner_id'])]

print("Missing Address IDs in database:{}".format(missing_addresses['address_id'].unique()))
print("Missing Owner IDs in database: {}".format(missing_owners['owner_id'].unique()))

Missing Address IDs in database:[]
Missing Owner IDs in database: ['4214' '4339' '5088' '5275' '6019' '7349' '7414' '8250' '8363' '8804'
 '8905' '952' '9524' '9715' '9749' '9942']


#### 2.5.2)📝 Data Quality Findings (Manual Audit)
During the initial integrity check between the listings dataset and the parent tables, the following inconsistencies were identified:

Missing Address IDs: [ ] (All addresses were successfully matched).

Missing Owner IDs: ['4214', '4339', '5088', '5275', '6019', '7349', '7414', '8250', '8363', '8804', '8905', '952', '9524', '9715', '9749', '9942']

Analysis: These 16 IDs existed in the listings file but were not present in the owners table. This confirms an inconsistency in the source data. As a corrective measure, these orphaned records will be excluded from the final migration to ensure 100% referential integrity in the MySQL database

#### 2.5.3) 🚀 Final Data Cleansing & Migration

In this section, I migrate the `listings` table. This table acts as a bridge between properties, their addresses, and their owners. To ensure database integrity and avoid Foreign Key errors, I normalize the IDs by converting them to strings, ensuring they match the VARCHAR format in the MySQL database.

**Strategy:** To fix **Error 1452**, I used a Filtering Strategy. I compared the listings DataFrame against the existing IDs in the owners and addresses tables. By using the .isin() method, I ensured that only 'valid' records were sent to the database, preventing the migration from crashing and maintaining 100% data integrity.

In [ ]:
# --- Table 4: LISTINGS ---

# 1. Path to the file
path_listings = r"C:\Users\vanes\OneDrive\Área de Trabalho\PROJETOS EM ANDAMENTO\PORTFÓLIO\PROJETO 2\Parte_2\csv\4-hospedagens.csv"

# 2. Load the CSV
df_listings = pd.read_csv(path_listings, encoding='utf-8')

# 3. Renaming columns to English (Prevents KeyError)
df_listings.columns = ['listing_id', 'property_type', 'address_id', 'owner_id', 'is_active']


# 4. Normalize IDs to String (to match VARCHAR in MySQL)
for col in ['listing_id', 'address_id', 'owner_id']:
    df_listings[col] = df_listings[col].astype(str).str.replace('.0', '', regex=False)

# 5. DIAGNOSIS & CORRECTION
# Get valid IDs from the parent tables already in MySQL
db_owners = pd.read_sql("SELECT owner_id FROM owners", con=engine)['owner_id'].tolist()
db_addresses = pd.read_sql("SELECT address_id FROM addresses", con=engine)['address_id'].tolist()

# 5.1. Filter the listings: Only keep rows where parent IDs exist in the database
df_listings_clean = df_listings[
    df_listings['owner_id'].isin(db_owners) & 
    df_listings['address_id'].isin(db_addresses)
].copy()

# 5.2. Calculate how many rows were removed for the report
removed_count = len(df_listings) - len(df_listings_clean)

# 6. Drop Duplicates
df_listings_clean = df_listings_clean.drop_duplicates(subset=['listing_id'])

# 7. Migration
# try:
#     df_listings_clean.to_sql('listings', con=engine, if_exists='append', index=False)
#     print(f"✅ Success! {len(df_listings_clean)} listings migrated.")
#     print(f"⚠️ Note: {removed_count} orphaned rows were excluded to maintain integrity.")
# except Exception as e:
#     print(f"❌ Error: {e}")



✅ Success! 9984 listings migrated.
⚠️ Note: 16 orphaned rows were excluded to maintain integrity.


#### 2.5.4) 🔏 Data Integrity Report: 
During the migration of the listings table, I identified 16 orphaned records (0.16% of the dataset) that referenced non-existent owners or addresses. To maintain Referential Integrity and satisfy database constraints, I implemented a filtering logic in Python to exclude these records, successfully migrating 9,984 consistent entries to the MySQL database.


In [29]:
df_listings_clean.head()

,listing_id,property_type,address_id,owner_id,is_active
0,1,casa,1,1,0
1,10,apartamento,10,1,0
2,100,hotel,100,100,0
3,1000,casa,1000,1000,1
4,10000,casa,10000,10000,0


#### 2.6) Processing Table: Bookings  

In this section, I process the `Bookings` table, which connects with two other tables: Customers and Listings. This ensures that every booking is linked to a valid customer and a specific property.

#### 2.6.1) Data Integrity Check - Part 1: Initial inspection of the Bookings dataset (source: `5-alugueis.csv`). This step verifies data encoding and ensures all columns are correctly renamed to match our English database schema.

In [9]:
# --- Table 5: BOOKINGS ---

# 1. Path to the file 
path = r"C:\Users\vanes\OneDrive\Área de Trabalho\PROJETOS EM ANDAMENTO\PORTFÓLIO\PROJETO 2\Parte_2\csv\5-alugueis.csv"

# 2. Load the CSV
df_bookings = pd.read_csv(path)

# 3. Renaming columns to match my English Database Schema 
df_bookings.columns = ["booking_id", "customer_id", "listing_id", "start_date", "end_date", "total_price"]

# 4. Displaying the first rows - Bookings table
df_bookings.head()

,booking_id,customer_id,listing_id,start_date,end_date,total_price
0,1,1,1,2022-08-12,2022-08-15,480.0
1,10,10,2,2022-11-04,2022-11-07,597.0
2,100,100,3,2022-11-22,2022-11-25,1404.0
3,1000,1000,9,2023-06-16,2023-06-21,995.0
4,10000,10000,66,2022-05-15,2022-05-17,1438.0


✅The Bookings dataset has passed initial integrity and encoding checks. Therefore, I can proceed with the Transformation and Load (TL) stages.

#### 2.6.2) Data Integrity Check - Part 2: Orphaned Records Discovery 

This step cross-references the Bookings dataset against existing Customers and Listings in the database. Initial results show a high volume of mismatched IDs. I am now investigating if this is a genuine data gap or a data type mismatch (Integer vs. String) before proceeding with the final filtering

In [10]:
# 1. Checking if all customers IDs in the Bookings CSV exist in the Customers table.
db_customers = pd.read_sql("SELECT customer_id FROM customers", con=engine)
missing_customers = df_bookings[~df_bookings['customer_id'].isin(db_customers['customer_id'])]



# 2. Checking if all listings IDs in the Bookings CSV exist in the Listings table.
db_listings = pd.read_sql("SELECT listing_id FROM listings", con=engine)
missing_listings = df_bookings[~df_bookings["listing_id"].isin(db_listings["listing_id"])]



print("Missing Customers IDs in database: {}".format(missing_customers['customer_id'].unique()))
print("Missing Listings IDs in database: {}".format(missing_listings["listing_id"].unique()))


Missing Customers IDs in database: [   1   10  100 ... 9997 9998 9999]
Missing Listings IDs in database: [ 1  2  3  9 66  6 11  5  7  8 12 13  4 10 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47
 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 67 68 69 70 71 72
 73 74 75 76 77]


#### 2.6.3) Data Cleansing - Part 1 - Solving Type Mismatch & Orphaned Records"

Initially, the filtering returned zero records because the IDs in the CSV were being interpreted as integers, while the database IDs were strings.<br> 
After converting the IDs to strings using .astype(str), the integrity check worked correctly."

* Original Records: 10,000

* Records After Cleansing: 9,983

* Findings: Only 17 orphaned records (0.17%) were identified and removed to maintain referential integrity.

In [11]:
# Cleansing Orphaned Records 
df_bookings_cleaned = df_bookings[df_bookings['listing_id'].isin(db_listings['listing_id'])]
df_bookings_cleaned = df_bookings_cleaned[df_bookings_cleaned['customer_id'].isin(db_customers['customer_id'])]

print("Original Records:{}".format(len(df_bookings)))
print("Records after cleansing: {}".format(len(df_bookings_cleaned)))

Original Records:10000
Records after cleansing: 0


In [ ]:
# Ensuring data type consistency by converting IDs to strings for comparison
df_bookings_cleaned = df_bookings[df_bookings['listing_id'].astype(str).isin(db_listings['listing_id'].astype(str))]

df_bookings_cleaned = df_bookings_cleaned[df_bookings_cleaned['customer_id'].astype(str).isin(db_customers['customer_id'].astype(str))]

print("The Bookings database has {} rows.".format(len(df_bookings_cleaned)))

The Bookings database has 9983 rows.


#### 2.6.4) Data Cleansing - Part 2 - Drop the duplicates
I implemented a custom validation script to check for duplicate booking_id entries. <br>
This ensures the dataset contains only unique transactions before the final load.

In [12]:
#Save data total before remove duplicates
# Salva o total antes de remover duplicatas
total_before = len(df_bookings_cleaned)

#Remove duplicates
# Remove as duplicatas
df_bookings_cleaned = df_bookings_cleaned.drop_duplicates(subset=["booking_id"])

#Checking if something has changed
# Verifica se algo mudou
total_after = len(df_bookings_cleaned)

if total_before > total_after:
    print(f"Heads up! {total_before - total_after} duplicate bookings were removed.")
else:
    print("No duplicates found. The data is unique!")

No duplicates found. The data is unique!


#### 2.6.5) Migration to MySQL 
In this final stage of the Bookings process, I ensured the data was perfectly formatted for the database environment. <br>
Key action included:

* Date Formatting: Converted the start_date and end_date columns to the proper datetime format using Pandas.

In [13]:
print(df_bookings.columns) 

Index(['booking_id', 'customer_id', 'listing_id', 'start_date', 'end_date',
       'total_price'],
      dtype='object')


In [ ]:
# Convert columns to datetime format
df_bookings_cleaned['start_date'] = pd.to_datetime(df_bookings_cleaned['start_date'])
df_bookings_cleaned['end_date'] = pd.to_datetime(df_bookings_cleaned['end_date'])


# Final Migration
# try:
#     df_bookings_cleaned.to_sql("bookings", con = engine, if_exists = "append", index=False)
#     print("✅ Table Bookings migrated successfully!")
# except Exception as e:
#     print("❌ Error migrating `Bookings: {}".format(e))

✅ Table Bookings migrated successfully!


#### 2.7) Processing Table: Reviews 
In this step, I will populate the `Reviews` table in the database. The data source is the `6-avaliacoes.csv` file

#### 2.7.1) Loading the `avaliacao` CSV and renaming the columns.
To maintain consistency across the project, I renamed the original columns to English: review_id, customer_id, listing_id, rating, and review_text.

In [ ]:
# --- Table 6: REVIEWS ---

# 1.Path the file 
path = r"C:\Users\vanes\OneDrive\Área de Trabalho\PROJETOS EM ANDAMENTO\PORTFÓLIO\PROJETO 2\Parte_2\csv\6-avaliacoes.csv"

# 2.Load the CSV 
df_reviews = pd.read_csv(path) 


# 3.Renaming the columns
df_reviews.columns = ['review_id', 'customer_id', 'listing_id', 'rating', 'review_text']


# 4.Displaying the first rows 
df_reviews.head()

,review_id,customer_id,listing_id,rating,review_text
0,1,1,1,1,Maravilhosa atendimento.\n
1,10,10,2,1,Maravilhosa infraestrutura.\n
2,100,100,3,1,Maravilhosa serviços.\n
3,1000,1000,9,5,Ótima infraestrutura.\n
4,10000,10000,66,5,Excelente infraestrutura.\n


#### 2.7.2) Reorder the columns to match with SQL structure. 
I reordered the columns to ensure the DataFrame structure aligns perfectly with the MySQL table schema, facilitating a seamless data injection.

In [59]:
# Reorder the columns 
df_reviews = df_reviews[['review_id', 'listing_id', 'customer_id', 'rating', 'review_text']]

# Displaying the first rows
df_reviews.head()

,review_id,listing_id,customer_id,rating,review_text
0,1,1,1,1,Maravilhosa atendimento.\n
1,10,2,10,1,Maravilhosa infraestrutura.\n
2,100,3,100,1,Maravilhosa serviços.\n
3,1000,9,1000,5,Ótima infraestrutura.\n
4,10000,66,10000,5,Excelente infraestrutura.\n


#### 2.7.3) Data Integrity Check: Orphaned Records Discovery 
In this step, I perform a cross-reference check to ensure every review is linked to an existing listing and customer in the database. Initial results showed a large number of 'missing' IDs. However, based on previous table experiences, this was identified as a data type mismatch rather than actual missing data.

In [60]:
# 1. Checking if all listings IDs in the Reviews CSV exist in the Listings table.
db_listings = pd.read_sql("SELECT listing_id FROM listings", con=engine)
missing_listings = df_reviews[~df_reviews['listing_id'].isin(db_listings['listing_id'])]



# 2. Checking if all customers IDs in the REviews CSV exist in the Customers table.
db_customers = pd.read_sql("SELECT customer_id FROM customers", con=engine)
missing_customers = df_reviews[~df_reviews["customer_id"].isin(db_customers["customer_id"])]



print("Missing Listings IDs in database: {}".format(missing_listings['listing_id'].unique()))
print("Missing Customers IDs in database: {}".format(missing_customers["customer_id"].unique()))


Missing Listings IDs in database: [ 1  2  3  9 66  6 11  5  7  8 12 13  4 10 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47
 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 67 68 69 70 71 72
 73 74 75 76 77]
Missing Customers IDs in database: [   1   10  100 ... 9997 9998 9999]


#### 2.7.4) Data Cleansing - Part 1 - Orphaned Records 
To resolve the mismatch, I converted all ID columns to strings using astype(str) before the comparison. This corrected the 'zero records' error and confirmed that the database actually contains 9,983 valid reviews. I then filtered the dataset to remove the 17 truly orphaned records that did not have a corresponding ID in the system.

In [61]:
# Cleansing Orphaned Records 
df_reviews_cleaned = df_reviews[df_reviews['listing_id'].isin(db_listings['listing_id'])]
df_reviews_cleaned = df_reviews_cleaned[df_reviews_cleaned['customer_id'].isin(db_customers['customer_id'])]


print("Original Records: {}".format(len(df_reviews)))
print("Records after cleansing: {}".format(len(df_reviews_cleaned)))




Original Records: 10000
Records after cleansing: 0


In [62]:
# Ensuring data type consistency by converting IDs to strings for comparison
df_reviews_cleaned = df_reviews[df_reviews['listing_id'].astype(str).isin(db_listings['listing_id'].astype(str))]

df_reviews_cleaned = df_reviews_cleaned[df_reviews_cleaned['customer_id'].astype(str).isin(db_customers['customer_id'].astype(str))]

print("The Reviews database has {} rows.".format(len(df_reviews_cleaned)))

The Reviews database has 9983 rows.


#### 2.7.5) Data Cleansing - Part 2 - Drop the duplicates
To ensure each review is unique and prevent redundant data in the database, I performed a check for duplicate records based on the review_id. No duplicates were found, confirming the integrity and uniqueness of the current dataset.

In [63]:
#Save data total before remove duplicates
# Salva o total antes de remover duplicatas
total_before = len(df_reviews_cleaned)

#Remove duplicates
# Remove as duplicatas
df_reviews_cleaned = df_reviews_cleaned.drop_duplicates(subset=["review_id"])

#Checking if something has changed
# Verifica se algo mudou
total_after = len(df_reviews_cleaned)

if total_before > total_after:
    print("Heads up! {} duplicate bookings were removed.".format(total_before - total_after))
else:
    print("No duplicates found. The data is unique!")

No duplicates found. The data is unique!


#### 2.7.6) Migration to MySQL 
In this final step, I performed the data injection into the MySQL database. After validating that the column order was correct and all data types were consistent, I used the to_sql method to persist the 9,983 cleaned records. The successful migration of the Reviews table marks the completion of the ETL process for the Insight Places project.

In [64]:
print(df_reviews.columns) 

Index(['review_id', 'listing_id', 'customer_id', 'rating', 'review_text'], dtype='object')


In [ ]:
# Final Migration
# try:
#     df_reviews_cleaned.to_sql("reviews", con = engine, if_exists = "append", index=False)
#     print("✅ Table Reviews migrated successfully!")
# except Exception as e:
#     print("❌ Error migrating `Reviews: {}".format(e))

✅ Table Reviews migrated successfully!


#### 🎯Conclusion and Next Steps
This first stage of the project focused on the `ETL (Extract, Transform, Load)` process, where I transformed raw data from CSV files into a structured and relational MySQL database.

Key achievements included:

* Data Consistency: Resolved critical data type mismatches (Integer vs. String) that would have caused data loss.

* Business Logic: Ensured referential integrity by filtering orphaned records across all six tables.

* Database Readiness: Successfully migrated nearly 10,000 clean records, ensuring the environment is ready for analysis.




🧱⛑️🔗Infrastructure, Security & Connectivity
A key pillar of this project was established in the initial setup of the database connection:

8 Professional Connectivity: I implemented the connection using SQLAlchemy with the PyMySQL driver, which is the industry standard for robust communication between Python and MySQL.

* Scalable Architecture: Using an engine object allows for efficient management of database resources and provides a secure foundation for the migration of all 6 tables.

* Data Protection: By structuring the migration via Python, I ensured that data was validated and cleaned (handling mismatches and duplicates) before ever reaching the production environment.

This technical foundation was crucial for the successful migration of the 9,983 records and will be the basis for the upcoming automation phase.

✨🔮What's next?<br>
Now that the database is populated and validated, the next phase will involve deep data analysis using SQL. Following that, I will return to Python to automate this entire pipeline, turning these manual steps into a scalable and efficient script.